src/Log/Engine/BaseLog.php

```diff
@@ -61,6 +69,22 @@ public function __construct(array $config = [])
         if (!empty($this->_config['types']) && empty($this->_config['levels'])) {
             $this->_config['levels'] = (array)$this->_config['types'];
         }
+
+        // Default to DefaultFormatter for BC with custom loggers that don't set `formatter` in $_defaultConfig
+        $formatterConfig = $this->_config['formatter'] ?? DefaultFormatter::class;
+        if (!is_array($formatterConfig)) {
+            $formatterConfig = ['className' => $formatterConfig];
+        }
+
+        $class = new ReflectionClass($formatterConfig['className']);
+        if (!$class->isSubclassOf(AbstractFormatter::class)) {
+            throw new InvalidArgumentException(sprintf(
+                'Formatter must extend `%s`, got `%s` instead',
+                AbstractFormatter::class,
+                $class->getName()
+            ));
+        }
+        $this->formatter = $class->newInstance($formatterConfig);
```

Comment:
Uses reflection instead of `instanceof` check makes it more difficult for static analyzers to deduce the class type :slightly_smiling_face:.

Reply:
psalm throws the same error on assignment with `new class-string()`. Have to tell it what's being instantiated.

Reply:
You would need the inline docblock to indicate what's the class for that `class-string`, no way of avoiding that.

Reply:
I think going back to `new` + instanceof makes that cleaner. If using ReflectionClass doesn't help psalm then it's just harder to annotate.